crates/forge/src/lockfile.rs

```diff
@@ -0,0 +1,403 @@
+                        // .gitmodules
+                        let maybe_branch = modules_with_branch.get(rel_path).map(|b| b.to_string());
+
+                        trace!(?maybe_branch, submodule = ?rel_path, "submodule branch");
+                        if let Some(branch) = maybe_branch {
+                            let dep_id = DepIdentifier::Branch {
+                                name: branch,
+                                rev: rev.to_string(),
+                                r#override: false,
+                            };
+                            e.insert(dep_id.clone());
+                            out_of_sync.insert(rel_path.to_path_buf(), dep_id);
+                            continue;
+                        }
+
+                        // Find out if rev has a tag associated with it.
+                        let maybe_tag =
+                            git.tag_for_commit(rev, &git.root.join(rel_path)).or_else(|err| {
+                                // Ignore Err: No such file or directory as it is possible that lib/
+                                // dir has been cleaned.
+                                if err.to_string().contains("No such file or directory") {
+                                    return Ok(None)
+                                }
+                                Err(err)
+                            })?;
+
+                        let dep_id = if let Some(tag) = maybe_tag {
+                            DepIdentifier::Tag {
+                                name: tag,
+                                rev: rev.to_string(),
+                                r#override: false,
+                            }
+                        } else {
+                            DepIdentifier::Rev { rev: rev.to_string(), r#override: false }
+                        };
+                        e.insert(dep_id.clone());
+                        out_of_sync.insert(rel_path.to_path_buf(), dep_id);
+                    }
+                }
+            }
+
+            return Ok(if out_of_sync.is_empty() { None } else { Some(out_of_sync) });
+        }
+
+        Ok(None)
+    }
+
+    /// Loads the lockfile from the project root.
+    ///
+    /// Throws an error if the lockfile does not exist.
+    pub fn read(&mut self) -> Result<()> {
+        if !self.lockfile_path.exists() {
+            return Err(eyre::eyre!("Lockfile not found at {}", self.lockfile_path.display()));
+        }
+
+        let lockfile_str = foundry_common::fs::read_to_string(&self.lockfile_path)?;
+
+        self.deps = serde_json::from_str(&lockfile_str)?;
+
+        trace!(lockfile = ?self.deps, "loaded lockfile");
+
+        Ok(())
+    }
+
+    /// Writes the lockfile to the project root.
+    pub fn write(&self) -> Result<()> {
+        foundry_common::fs::write_json_file(&self.lockfile_path, &self.deps)?;
+        trace!(at= ?self.lockfile_path, "wrote lockfile");
+
+        Ok(())
+    }
```

Comment:
Would be nice if the lockfile was rendered as pretty formatted JSON, makes it easier to verify the `rev`'s